RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1628 @@
+    auto& clusterEnergy = alpaka::declareSharedVar<float[400], __COUNTER__>(acc);
+    auto& rhFracSum = alpaka::declareSharedVar<float[1500], __COUNTER__>(acc);
+    auto& seeds = alpaka::declareSharedVar<int[400], __COUNTER__>(acc);
+    auto& rechits = alpaka::declareSharedVar<int[1500], __COUNTER__>(acc);
+
+    if (once_per_block(acc)) {
+      nRHNotSeed = nRHTopo - nSeeds + 1;  // 1 + (# rechits per topoId that are NOT seeds)
+      topoSeedBegin = pfClusteringVars[topoId].topoSeedOffsets();
+      tol = pfClusParams.stoppingTolerance() *  // stopping tolerance * tolerance scaling
+            powf(fmaxf(1.0, nSeeds - 1.0), 2.0);
+      gridStride = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+      iter = 0;
+      notDone = true;
+
+      int i = pfClusteringVars[topoSeedBegin].topoSeedList();
+      if (pfRecHits[i].layer() == PFLayer::HCAL_BARREL1)
+        rhENormInv = pfClusParams.recHitEnergyNormInvEB_vec()[pfRecHits[i].depth() - 1];
+      else if (pfRecHits[i].layer() == PFLayer::HCAL_ENDCAP)
+        rhENormInv = pfClusParams.recHitEnergyNormInvEE_vec()[pfRecHits[i].depth() - 1];
+      else
+        printf("Rechit %d has invalid layer %d!\n", i, pfRecHits[i].layer());
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    for (int n = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]; n < nRHTopo; n += gridStride) {
+      if (n < nSeeds)
+        seeds[n] = pfClusteringVars[topoSeedBegin + n].topoSeedList();
+      if (n < nRHNotSeed - 1)
+        rechits[n] =
+            fracView[pfClusteringVars[pfClusteringVars[topoSeedBegin].topoSeedList()].seedFracOffsets() + n + 1]
+                .pfrhIdx();
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    if constexpr (debug) {
+      if (once_per_block(acc)) {
+        printf("\n===========================================================================================\n");
+        printf("Processing topo cluster %d with nSeeds = %d nRHTopo = %d and seeds (", topoId, nSeeds, nRHTopo);
+        for (int s = 0; s < nSeeds; s++) {
+          if (s != 0)
+            printf(", ");
+          printf("%d", getSeedRhIdx(seeds, s));
+        }
+        if (nRHTopo == nSeeds) {
+          printf(")\n\n");
+        } else {
+          printf(") and other rechits (");
+          for (int r = 1; r < nRHNotSeed; r++) {
+            if (r != 1)
+              printf(", ");
+            if (r <= 0) {
+              printf("Invalid rhNum (%d) for get RhFracIdx!\n", r);
+            }
+            printf("%d", rechits[r - 1]);
+          }
+          printf(")\n\n");
+        }
+      }
+      alpaka::syncBlockThreads(acc);  // all (or none) threads call sync
+    }
+
+    // Set initial cluster position (energy) to seed rechit position (energy)
+    for (int s = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]; s < nSeeds; s += gridStride) {
```

Comment:
I guess it could be useful to introduce a new class, `block_elements_with_stride`, to be used as
```suggestion
    for (int s: block_elements_with_stride(acc, nSeeds) {
```
?

Suggestions for a better name are welcome.

Reply:
I think this would be useful